### ISSUE TITLE
Adjusting the Master Pitch range to above/below 12 semitones

### ISSUE THREAD

**Issue by jshamble** [LMMS/lmms#5016]
Can there be a way to increase/decrease the range of the master pitch slider to above 12 semitones and below 12 semitones to +/- N semitones (where N is user-defined, similar to how adjusting the pitch-bend slider works).

**Comment by thmueller64** [comment]
I would like to work on this, if it is okay.

**Comment by thmueller64** [comment]
Here is my suggestion for the basic layout. I wanted to follow the design of the Sf2 Player mentioned in the issue. What do you think about the layout? I did some basic testing by adjusting the range and the master pitch during playback and, from what I can tell, it is behaving as expected. Next, I'll look at the behavior when automating the master pitch and adjusting the range simultaneously. When this is done, a way to save and load the new parameters has to be implemented.

![master_pitch_demo](https://user-images.githubusercontent.com/64359888/80748778-c26a0400-8b25-11ea-89fa-5c6bceb65d30.gif)

I tried to structure/track the requirements and possible issues. Please feel free to discuss the approach:
- [ ] Basic layout of the range adjustment field 
- [x] Test manual adjustment of master pitch and -range during midi playback
- [ ] Automation range must be able to adjust dynamically (--> what should happen when range is reduced, but automation exceeds the new threshold?)
- [ ] Ability to save and load projects using this option, considering possible implications (e.g., what happens when a project using the feature is imported with an older version of lmms?)

**Comment by jshamble** [comment]
@thmueller64 Looks good so far, thanks for picking this up!

In regards to the layout is seems fine to me, although it kind of stands out as it's on its own, perhaps to rectify this and balance the scales we can a similar range selector to the master volume slider to the left of it as well, I've opened a PR for that as well: https://github.com/LMMS/lmms/issues/5479

------

At a deeper level I would also think why is there no way for the user to manually edit the range of values in the automation editor? I would think that would be a bit more involved though, as perhaps it would require some sanitation of the user input values, I see there is a similar PR https://github.com/LMMS/lmms/issues/4563 that has been opened for that already though.

------

In regards to the functionality: when the range is reduced, I would follow the suit of how automation functions for the pitch bend slider in the soundfont player as shown in the original post. 

1) I add the pitch bend to the automation editor by [ctrl + drag and dropping the slider] set the range to 10, add in a random automation value

![1](https://user-images.githubusercontent.com/6364311/80764850-705dba00-8af6-11ea-8561-f854dded0064.PNG)

2) I reduce the range, and re-open the automation editor, it now displays a new "clamped" range.

![2](https://user-images.githubusercontent.com/6364311/80764866-75bb0480-8af6-11ea-9275-708546fb1dda.PNG)

3) however, If I then change the range back, I see that the "original value" that was not visible in the automation editor was still preserved, even though it exceeded the clamping threshold.

![1](https://user-images.githubusercontent.com/6364311/80764870-79e72200-8af6-11ea-989c-9d30afb319bd.PNG)

In this regard, I would think it is best to follow suit, i.e. preserve the original value of the master pitch even though the range is clamped **until it is manually changed by the user**

**Comment by thmueller64** [comment]
Thanks @jshamble for your input.

> In regards to the layout is seems fine to me, although it kind of stands out as it's on its own, ...

You're right, it does not really fit this way.

>...perhaps to rectify this and balance the scales we can a similar range selector to the master volume slider to the left of it as well, I've opened a PR for that as well: #5479

I get the idea, however, this could probably overload the UI with an option that is not very likely to be used. Thanks for opening #5479, it would be nice to get more opinions on this. 

> In regards to the functionality: when the range is reduced, I would follow the suit of how automation functions for the pitch bend slider in the soundfont player as shown in the original post.

I followed your steps for the pitch range in sf player. Regarding automation, the clamping-but-preserving behavior is the same for the master pitch range.

A rather fundamental question to this issue:
The main advantage of the "range" widget in soundfont player is that you can quickly swipe over large, fine grained, ranges, e.g. from [-2000, 2000] cents, right? Since the master pitch slider is in the small and coarse scale of semitones, I'm not really sure whether a "range" widget would make sense entirely. 
This favors the automation editor solution you mention, switching to cents for the master pitch or simply extending the semitone range to e.g. [-64, 64].

**Comment by jshamble** [comment]
@thmueller64 Yes, I do agree with that - but I still think that this solution is better than nothing - I would rather have a solution rather than no solution. If agreed, I'd like to push this for the next release, and perhaps come up with an improved user experience widget design in another open issue.

**Comment by thmueller64** [comment]
Sorry @jshamble for the delay, I totally missed your comment on this! I'm fine with that, although a second opinion on the topic would be great before working on a pull request.

**Comment by jshamble** [comment]
@thmueller64 It doesn't seem like anyone is chiming in - I would still like to see this feature though! I've been using LMMS for 12+ years and I still plan on using it for a long time, this feature would make a pretty big impact on me (and I'm sure a lot of other people as well)!

**Comment by thmueller64** [comment]
Hello @jshamble , you can check out a build including the changes under the PR linked above.

**Comment by allejok96** [comment]
Why not just throw away the slider and let the master pitch be a LCD? Then we have fine grained control, clear visual feedback and almost no limitations.

**Comment by bauhaushousedog** [comment]
is this an easy first issue to work if i don't have experience with the codebase? (i have some knowledge on c++)

**Comment by zonkmachine** [comment]
> is this an easy first issue to work if i don't have experience with the codebase? (i have some knowledge on c++)

Yes, but it looks like @thmueller64 already has a fix for this here: https://github.com/LMMS/lmms/pull/6262
Maybe you could start by building and testing this?
